### PR TITLE
Create Image Tools Interface

### PR DESCRIPTION
**Pull request by RobDolinMS**
Describes an interface that the image-tools project could provide for image validation

Signed-off-by: Rob Dolin <robdolin@microsoft.com>

### CODE REVIEW

**Comment by wking** [comment]
On Mon, Mar 20, 2017 at 02:48:12PM -0700, Rob Dolin (MSFT) wrote:
> Describes an interface that the image-tools project could provide
> for image validation

Maybe this should be an image-tools issue/PR?  I expect a certifier to
be using image-tools directly, and don't see a need to specify its API
independent of the implementation.  For comparison, the runtime
command line API in opencontainers/runtime-tools#321 is aimed at the
*runtime* API, not at the runtime-tools API.

For your UI with number-of-cases, failed cases, verbose listing, one
approach would be to make image-tools's image validator a TAP
producer.  Then you could run it from a TAP harness to get a compact
or verbose report.  For an example of this in practice, see the
comments in opencontainers/runtime-tools#308.  Prove [1] is one such
harness and has options for:

   -Q,  --QUIET           Only print summary results.
   -f,  --failures        Show failed tests.
   -o,  --comments        Show comments.

If we get granular TAP results like opencontainers/runtime-tools#308
(or in any OCI compliance testing), I think the TAP comments are a
great place to put the spec links you're looking for.

And you could probably get all of this pretty handling with a number
of test harnesses or suites that output structured results.  I'm
floating TAP because I know it already handles these cases, and runC
(via bats [2]) and runtime-tools (opencontainers/runtime-tools#302)
are already using some TAP.  But if OCI maintainers prefer other
structured test-result formats, they would be fine too.

[1]: http://perldoc.perl.org/prove.html
[2]: https://github.com/sstephenson/bats/tree/v0.4.0#bats-bash-automated-testing-system

**Comment by caniszczyk** [comment]
cc: @opencontainers/image-tools-maintainers maintainers and @opencontainers/image-spec-maintainers for RFC

**Comment by caniszczyk** [comment]
@RobDolinMS is it possible for you to bring this up at the OCI Dev meeting and gather feedback from the OCI TDC?

**Comment by caniszczyk** [comment]
@RobDolinMS also any thoughts about just merging this into the README instead of a separate doc? (just keeps it easier to track)

**Comment by RobDolinMS** [comment]
@caniszczyk Yes, I mentioned this PR on the OCI Dev ConCall a few hours ago.

I'm also tagging @opencontainers/image-spec-maintainers and @opencontainers/image-tools-maintainers to please take a look at this DRAFT (PR #26) requested interface from the @opencontainers/certification-maintainers: https://github.com/opencontainers/certification/pull/26/files

**Comment by RobDolinMS** [comment]
@opencontainers/image-tools-maintainers - Does this proposal roughly align to how the code runs?

**Comment by wking** [comment]
On Mon, Apr 03, 2017 at 04:45:04PM -0700, Rob Dolin (MSFT) wrote:
> Does this proposal roughly align to how the code runs?

Current invocation is documented here [1].

[1]: https://github.com/opencontainers/image-tools/blob/master/man/oci-image-tool-validate.1.md#examples

**Comment by vbatts** [comment]
this tab has been open for a few days. I like the direction, just want to ensure it plumbs-up to expected behavior

File: image-tools-interface.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+Inputs:
+* (REQUIRED): Path or reference to a container image
+* (OPTIONAL): If the user would like verbose output
+
+Outputs:
+* Number of cases run
+* Number of cases passed
+* Number of cases failed
+* List of cases failed
+* Version of OCI Image Tools
```
Comment (vbatts) [inline image-tools-interface.md]: and maybe git commit of the tool? 
and version of the OCI spec tested against.
And the timestamp the validation just ran.
Also, presumably this would need to output the digests of the objects tested.

With this data points in a structured output (i.e. xml, json, etc) then it could be more deterministically used for badges, etc.

Comment (wking) [inline image-tools-interface.md]: > and maybe git commit of the tool?

*I* like this, but note that in a similar context the OCI consensus (as voiced by @duglin [here][1]) was that a consistent API for extracting the implementation version was not required for interop.

> And the timestamp the validation just ran.

Does this matter?  Hopefully the version of the test suite/tools and the digest(s) of the tested object would be sufficient without recourse to the validation timestamp.  Higher-level tooling can always add a timestamp if it feels it would be useful.

> Also, presumably this would need to output the digests of the objects tested.

This is tricky for stuff outside of CAS.  For example, if validation was pointed at an image-layout via a HTTP URL, or at a directory, what digest hash do you show for complaining about “I couldn't find `index.json` there”?

[1]: http://ircbot.wl.linuxfoundation.org/meetings/opencontainers/2016/opencontainers.2016-07-20-21.03.log.html#l-75

**Comment by vbatts** [comment]
for a deterministic badge of validation, I think those three things are
needed/suitable to triangulate precisely _what_ was tested on _which_
artifacts, and a timestamp for freshness check.

**Comment by wking** [comment]
On Wed, Apr 05, 2017 at 12:48:00PM -0700, Vincent Batts wrote:
> … and a timestamp for freshness check.

But what is the purpose of the freshness check?  I think the tooling
for deprecating old versions [1] (and their certification results)
should depend on the tested OCI version, not on the timestamp of the
testing.

[1]: https://github.com/opencontainers/certification/blame/76149742708a6eedef64c683bfb5f889d15e4b11/README.md#L89-L90

**Comment by vbatts** [comment]
Perhaps timestamp isn't a hard requirement, but rather a nice-to-have.

File: image-tools-interface.md (hunk 1 above)
Comment (RobDolinMS) [inline image-tools-interface.md]: Replying to comments from @vbatts and @wking:

>and maybe git commit of the tool?

I was thinking that 'version of the OCI Image Tools' would be sufficient since only released versions of the tools will be relevant for image validation.  

>and version of the OCI spec tested against.

I didn't think of this, but it makes sense.  I'll update the PR.

>And the timestamp the validation just ran.
I'm not sure this is needed since a released version of the OCI Image Tool, a released version of the OCI Image Spec, and the target image should have the same output every time.

>Also, presumably this would need to output the digests of the objects tested.

@vbatts Would you mind suggesting what this might look like?

Comment (RobDolinMS) [inline image-tools-interface.md]: >I didn't think of [version of the OCI spec tested against], but it makes sense. I'll update the PR.

PR is updated at lines 11, 12, 30, and 40

File: image-tools-interface.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# Image Tools Interface
+
+The below document describes an interface for the OCI Image Tools project as desired by the OCI Certification Program working group (Cert WG.)
+
+Goals:
+* An engineer can verify a container image complies with the OCI Image Spec (https://github.com/opencontainers/image-spec/releases)
```
Comment (wking) [inline image-tools-interface.md]: Trailing whitespace on this line (and a few others):

```
$ git diff --check origin/master..origin/pr/26
image-tools-interface.md:6: trailing whitespace.
+* An engineer can verify a container image complies with the OCI Image Spec (https://github.com/opencontainers/image-spec/releases) 
image-tools-interface.md:11: trailing whitespace.
+* (OPTIONAL): Which version of the OCI Image Spec to validate against 
image-tools-interface.md:25: trailing whitespace.
+INPUT: 
image-tools-interface.md:28: trailing whitespace.
+OUTPUT (default): 
image-tools-interface.md:31: trailing whitespace.
+Case 12: FAILED: Image using reserved field key: https://github.com/opencontainers/image-spec/blob/master/descriptor.md#reserved 
image-tools-interface.md:32: trailing whitespace.
+Results: 
image-tools-interface.md:48: trailing whitespace.
+Results: 
```

File: image-tools-interface.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+The below document describes an interface for the OCI Image Tools project as desired by the OCI Certification Program working group (Cert WG.)
+
+Goals:
+* An engineer can verify a container image complies with the OCI Image Spec (https://github.com/opencontainers/image-spec/releases) 
+
+Inputs:
+* (REQUIRED): Path or reference to a container image
+* (OPTIONAL): If the user would like verbose output
+* (OPTIONAL): Which version of the OCI Image Spec to validate against 
+(IF the OCI Image Tool supports multiple OCI Image Spec versions; Default to highest supported)
```
Comment (wking) [inline image-tools-interface.md]: “IF” → “If”?

File: image-tools-interface.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+Case 12: FAILED: Image using reserved field key: https://github.com/opencontainers/image-spec/blob/master/descriptor.md#reserved 
+Results: 
+* Ran: 24 / 24 (100%) validation cases
+* Passed: 23 / 24 (96%)
+* Failed: 1 / 24 (4%)
+```
+
+OUTPUT (verbose):
+```
+OCI Image Tools vX.Y.Z validating <image> with OCI Image Spec vA.B.C:
```
Comment (stevvooe) [inline image-tools-interface.md]: It might be good to look at [rust's compiler error documentation](https://doc.rust-lang.org/error-index.html) for the items that might be required here for each error. As this is, it is a little hard to parse.

Let's also take a peak at Go's output:

```
=== RUN   TestCopyDirectory
--- PASS: TestCopyDirectory (0.00s)
=== RUN   TestSimpleDiff
--- PASS: TestSimpleDiff (0.00s)
=== RUN   TestDirectoryReplace
--- PASS: TestDirectoryReplace (0.00s)
=== RUN   TestRemoveDirectoryTree
--- PASS: TestRemoveDirectoryTree (0.00s)
=== RUN   TestFileReplace
--- PASS: TestFileReplace (0.00s)
=== RUN   TestUpdateWithSameTime
--- PASS: TestUpdateWithSameTime (0.00s)
=== RUN   TestBaseDirectoryChanges
--- PASS: TestBaseDirectoryChanges (0.00s)
PASS
ok  	github.com/containerd/containerd/fs	0.007s
```

Each line is marked as "starting test" and "ending test" (`===` and `---` respectively), the we get a parseable package name at the end. Timing is less important for validation, so we don't need both, but we need a way to differentiate a result from other information.

Something like this might work:

```
* TestName1: PASSED
* TestNameFailure: FAILED https://github.com/opencontainers/image-spec/blob/master/descriptor.md#reserved
... Information about failure.
* TestName2: PASSED
# Results: 23 / 24 (96%) passed
```

Notice that the status and result lines are marked with a leading character, leaving the other output to be expressive. There may be better characters than `*` for marking these lines, but we can see how we mark what is consumable by a leading character.

**Comment by caniszczyk** [comment]
Any comments here @stephenrwalli

File: image-tools-interface.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
+# Image Tools Interface
+
+The below document describes an interface for the OCI Image Tools project as desired by the OCI Certification Program working group (Cert WG.)
+
+Goals:
+* An engineer can verify a container image complies with the OCI Image Spec (https://github.com/opencontainers/image-spec/releases) 
+
+Inputs:
+* (REQUIRED): Path or reference to a container image
```
Comment (alban) [inline image-tools-interface.md]: If distribution is out of scope of image-spec, what would a "reference" be? A `https://` URL to the OCI Image Layout directory? And the software would download `${URL}/index.json` and other files as required by the [image layout content](https://github.com/opencontainers/image-spec/blob/master/image-layout.md#content)?

File: image-tools-interface.md (hunk 4 above)
Comment (cyphar) [reply image-tools-interface.md]: We could also use [TAP](https://en.wikipedia.org/wiki/Test_Anything_Protocol) rather than making our own output format.
